torchgeo/datasets/openstreetmap.py

```diff
@@ -0,0 +1,555 @@
+"""OpenStreetMap dataset."""
+
+import contextlib
+import hashlib
+import json
+import pathlib
+import re
+import time
+import warnings
+from collections.abc import Callable, Iterable
+from typing import Any, ClassVar, cast
+from urllib.parse import urlencode
+from urllib.request import Request, urlopen
+
+import geopandas as gpd
+import matplotlib.patches as mpatches
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import shapely
+from geopandas import GeoDataFrame
+from matplotlib.figure import Figure
+from pyproj import CRS
+
+from .errors import DatasetNotFoundError
+from .geo import VectorDataset
+from .utils import GeoSlice, Path
+
+
+class OpenStreetMap(VectorDataset):
+    """OpenStreetMap dataset.
+
+    The `OpenStreetMap <https://www.openstreetmap.org/>`__ dataset provides
+    access to crowd-sourced geographic data. This implementation uses the
+    `Overpass API <https://wiki.openstreetmap.org/wiki/Overpass_API>`__
+    to query and download OSM data for a specified geographic bounding box
+    at initialization, then allows spatial querying of the cached data.
+
+    Dataset features
+    ----------------
+
+    * Vector data (points, lines, polygons) for various geographic features
+    * Flexible querying by class configuration (buildings, highways, amenities, etc.)
+    * Data fetched once at initialization and cached locally
+    * Class-based labeling with priority-based assignment
+
+    Class priority and label assignment
+    ------------------------------------
+    `classes` is a list of dicts defining feature classes. Each has `name` (str) and `selector` (list of OSM tag filters).
+    Features are assigned labels based on the order of classes in this list:
+
+    - First class gets label=1, second gets label=2, etc.
+    - If a feature matches multiple classes, it receives the label of the first matching class
+    - Features that don't match any class get label=0 (background)
+
+    Example::
+
+        classes = [
+            {'name': 'buildings', 'selector': [{'building': '*'}]},        # label=1
+            {'name': 'roads', 'selector': [{'highway': '*'}]},             # label=2
+            {'name': 'commercial', 'selector': [{'landuse': 'commercial'}]} # label=3
+        ]
+
+        # A feature with tags {'building': 'yes', 'landuse': 'commercial'}
+        # would get label=1 (buildings) because buildings comes first
+
+    If you use this dataset in your research, please cite the following source:
+
+    * https://www.openstreetmap.org/copyright
+
+    .. versionadded:: 0.8
+    """
+
+    _overpass_endpoints: ClassVar[list[str]] = [
+        'https://overpass-api.de/api/interpreter',
+        'https://overpass.kumi.systems/api/interpreter',
+    ]
+
+    _min_request_interval = 1.0
+    _last_request_time = 0.0
+
+    def __init__(
+        self,
+        bbox: tuple[float, float, float, float],
+        classes: list[dict[str, Any]],
+        paths: Path | Iterable[Path] = 'data',
```

Comment:
It seems like `Iterable[Path]` isn't actually supported. We can remove this if we don't want to support it. I think mypy is okay with this.